### ISSUE TITLE
Move from C++11 to C++17 incompatible with latest OpenFst (i.e., 1.7.5)

### ISSUE THREAD

**Issue by PierreAndreNoel** [kylebgorman/pynini#19]
Version 2.0.9.post2 [changed](https://github.com/kylebgorman/pynini/commit/76178c3e25d6570b425b2a30f188cf835f7a6f23#diff-2eeaed663bd0d25b7e608891384b7298L24-R24) one of gcc's hard-coded flags from `-std=c++11` to `-std=c++17`. However, the latest OpenFst (i.e., 1.7.5) still uses the hard-coded `-std=c++11` flag, which induces incompatible symbols in the `*.so`.

Example:
```Dockerfile
FROM python:3.7-buster


# OpenFst

ENV FST_VERSION "1.7.5"
ENV FST_DOWNLOAD_PREFIX "http://www.openfst.org/twiki/pub/FST/FstDownload"

RUN cd /tmp \
    && wget -q ${FST_DOWNLOAD_PREFIX}/openfst-${FST_VERSION}.tar.gz \
    && tar -zxf openfst-${FST_VERSION}.tar.gz \
    && rm openfst-${FST_VERSION}.tar.gz

RUN cd /tmp/openfst-${FST_VERSION} \
    # && mv configure configure.bad \  # Uncomment this and following for workaround
    # && cat configure.bad | sed 's/-std=c++11/-std=c++17/' > configure \
    # && rm configure.bad \
    # && chmod +x configure \
    && ./configure  --enable-grm \
    && make --jobs=4 \
    && make install \
    && rm -rd /tmp/openfst-${FST_VERSION}

RUN ldconfig


# Pynini

ENV PYNINI_VERSION "2.0.9.post2"
ENV PYNINI_DOWNLOAD_PREFIX "http://www.opengrm.org/twiki/pub/GRM/PyniniDownload"

RUN cd /tmp \
    && wget -q ${PYNINI_DOWNLOAD_PREFIX}/pynini-${PYNINI_VERSION}.tar.gz \
    && tar -zxf pynini-${PYNINI_VERSION}.tar.gz \
    && rm pynini-${PYNINI_VERSION}.tar.gz

RUN mkdir /tools \
    && mv /tmp/pynini-${PYNINI_VERSION} /tools/pynini \
    && cd /tools/pynini \
    && python setup.py install


# Import test
RUN python -c "import pynini"
```

From a folder containing the above `Dockerfile`, calling `docker build .` fails on the last line with the error message:
```
Step 11/11 : RUN python -c "import pynini"
 ---> Running in 369a30cc34d8
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: /usr/local/lib/python3.7/site-packages/pynini-2.0.9.post2-py3.7-linux-x86_64.egg/pynini.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _ZNK3fst8internal14DenseSymbolMap4FindESt17basic_string_viewIcSt11char_traitsIcEE
The command '/bin/sh -c python -c "import pynini"' returned a non-zero code: 1
```
In short, Pynini expects the symbol `fst::internal::DenseSymbolMap::Find(std::basic_string_view<char, std::char_traits<char> >) const` but the OpenFst `.so` instead has `fst::internal::DenseSymbolMap::Find(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const`.

**Comment by kylebgorman** [comment]
Thanks for the report.

I can fix the compilation flags used for OpenFst upstream (not Pynini
though), but cannot replicate the issue locally. (I didn't try the Docker
version.)

**Comment by PierreAndreNoel** [comment]
Are you on mac? I believe that OSX's gcc links to clang, hence the different behavior.

**Comment by kylebgorman** [comment]
Nope, I don't have any easy way to test on Mac, except by writing recipes
for Conda-Forge. (That said the current state of things builds for Mac on
conda-forge and in fact unless you have an allergy to conda---some people
don't care for it, I'm sure---I'd strongly recommend using it to install
this.)

**Comment by PierreAndreNoel** [comment]
To clarify, I'm not on mac. My Dockerfile is based on Debian Buster, and I reproduced the problem on Ubuntu Bionic (outside docker).

**Comment by kylebgorman** [comment]
No worries, I don't doubt you, just can't reproduce myself from Bionic.

**Comment by lumpidu** [comment]
I have run the above Dockerfile with updated release-versions: pynini 2.1.1 & openfst 1.7.7 and I get different errors:

First, I need to add an additional build step so that Cython is found. But even then, I get the same error as the issue author.

Here is my complete Dockerfile:

```Dockerfile
FROM python:3.7-buster

# OpenFst

ENV FST_VERSION "1.7.7"
ENV FST_DOWNLOAD_PREFIX "http://www.openfst.org/twiki/pub/FST/FstDownload"

RUN cd /tmp \
    && wget -q ${FST_DOWNLOAD_PREFIX}/openfst-${FST_VERSION}.tar.gz \
    && tar -zxf openfst-${FST_VERSION}.tar.gz \
    && rm openfst-${FST_VERSION}.tar.gz

RUN cd /tmp/openfst-${FST_VERSION} \
    # && mv configure configure.bad \  # Uncomment this and following for workaround
    # && cat configure.bad | sed 's/-std=c++11/-std=c++17/' > configure \
    # && rm configure.bad \
    # && chmod +x configure \
    && ./configure  --enable-grm \
    && make --jobs=4 \
    && make install \
    && rm -rd /tmp/openfst-${FST_VERSION}

RUN ldconfig


# Pynini

ENV PYNINI_VERSION "2.1.1"
ENV PYNINI_DOWNLOAD_PREFIX "http://www.opengrm.org/twiki/pub/GRM/PyniniDownload"

RUN cd /tmp \
    && wget -q ${PYNINI_DOWNLOAD_PREFIX}/pynini-${PYNINI_VERSION}.tar.gz \
    && tar -zxf pynini-${PYNINI_VERSION}.tar.gz \
    && rm pynini-${PYNINI_VERSION}.tar.gz

RUN mkdir /tools \
    && mv /tmp/pynini-${PYNINI_VERSION} /tools/pynini \
    && cd /tools/pynini \
    && python3 -m pip install --upgrade cython \
    && python3 setup.py install

# Import test
RUN python -c "import pynini"
```
it would be really helpful, if this project would add a github actions process, so that one could exactly reproduce the build steps necessary. For me, neither Mac OS X build nor the Docker-based approach work.

**Comment by PierreAndreNoel** [comment]
@lumpidu It works for me if you uncomment my fix.
```
FROM python:3.7-buster

# OpenFst

ENV FST_VERSION "1.7.7"
ENV FST_DOWNLOAD_PREFIX "http://www.openfst.org/twiki/pub/FST/FstDownload"

RUN cd /tmp \
    && wget -q ${FST_DOWNLOAD_PREFIX}/openfst-${FST_VERSION}.tar.gz \
    && tar -zxf openfst-${FST_VERSION}.tar.gz \
    && rm openfst-${FST_VERSION}.tar.gz

RUN cd /tmp/openfst-${FST_VERSION} \
    && mv configure configure.bad \
    && cat configure.bad | sed 's/-std=c++11/-std=c++17/' > configure \
    && rm configure.bad \
    && chmod +x configure \
    && ./configure  --enable-grm \
    && make --jobs=4 \
    && make install \
    && rm -rd /tmp/openfst-${FST_VERSION}

RUN ldconfig


# Pynini

ENV PYNINI_VERSION "2.1.1"
ENV PYNINI_DOWNLOAD_PREFIX "http://www.opengrm.org/twiki/pub/GRM/PyniniDownload"

RUN cd /tmp \
    && wget -q ${PYNINI_DOWNLOAD_PREFIX}/pynini-${PYNINI_VERSION}.tar.gz \
    && tar -zxf pynini-${PYNINI_VERSION}.tar.gz \
    && rm pynini-${PYNINI_VERSION}.tar.gz

RUN mkdir /tools \
    && mv /tmp/pynini-${PYNINI_VERSION} /tools/pynini \
    && cd /tools/pynini \
    && python3 -m pip install --upgrade cython \
    && python3 setup.py install

# Run pynini's tests
RUN cd /tools/pynini && python pynini_test.py -v -f
```

What we ended up doing is that I made `manylinux2014` Python wheels that I pushed to our local PyPI. OpenFst's `.so` are all bundled in the wheel, so a simple `pip`/`poetry` install gets you a working `pynini`. @kylebgorman please let me know if this would have some interest to you.

**Comment by kylebgorman** [comment]
Yes, I am not knowledgeable here about how that works but I would love to
incorporate it if you can assist.

I also should note these assets are all available from Conda-Forge
nowadays: ‘conda install -c conda-forge pynini’ should work out of the box.

**Comment by PierreAndreNoel** [comment]
> Yes, I am not knowledgeable here about how that works but I would love to incorporate it if you can assist.

I've got the OK to share our code for that. I will adapt it a bit and come back to you soon.


> I also should note these assets are all available from Conda-Forge nowadays: ‘conda install -c conda-forge pynini’ should work out of the box.

Not everyone can use Conda-Forge. In the status quo, if the policy is "all dependencies must be `pip`/`poetry`-installable", then I either can't use `pynini`, or I must provide my own wheels for it.

**Comment by kylebgorman** [comment]
Understood re: Conda. It's not something I'm religious about, just the
first packaging solution that worked for me when there are non-Python
dependencies afoot. We'd welcome additional packaging solutions.
